src/set_theory/ordinal/natural_ops.lean

```diff
@@ -0,0 +1,366 @@
+instance : has_one nat_ordinal := ⟨to_nat_ordinal 1⟩
+instance : has_well_founded nat_ordinal := ordinal.has_well_founded
+instance : linear_order nat_ordinal := ordinal.linear_order
+
+@[simp] theorem to_nat_ordinal_zero : to_nat_ordinal 0 = 0 := rfl
+@[simp] theorem of_nat_ordinal_zero : of_nat_ordinal 0 = 0 := rfl
+@[simp] theorem to_nat_ordinal_one : to_nat_ordinal 1 = 1 := rfl
+@[simp] theorem of_nat_ordinal_one : of_nat_ordinal 1 = 1 := rfl
+
+@[simp] theorem to_nat_ordinal_eq_zero (a) : to_nat_ordinal a = 0 ↔ a = 0 := iff.rfl
+@[simp] theorem of_nat_ordinal_eq_zero (a) : of_nat_ordinal a = 0 ↔ a = 0 := iff.rfl
+@[simp] theorem to_nat_ordinal_eq_one (a) : to_nat_ordinal a = 1 ↔ a = 1 := iff.rfl
+@[simp] theorem of_nat_ordinal_eq_one (a) : of_nat_ordinal a = 1 ↔ a = 1 := iff.rfl
+
+@[simp] theorem to_nat_ordinal_lt_iff {a b} : to_nat_ordinal a < to_nat_ordinal b ↔ a < b := iff.rfl
+@[simp] theorem to_nat_ordinal_le_iff {a b} : to_nat_ordinal a ≤ to_nat_ordinal b ↔ a ≤ b := iff.rfl
+@[simp] theorem of_nat_ordinal_lt_iff {a b} : of_nat_ordinal a < of_nat_ordinal b ↔ a < b := iff.rfl
+@[simp] theorem of_nat_ordinal_le_iff {a b} : of_nat_ordinal a ≤ of_nat_ordinal b ↔ a ≤ b := iff.rfl
+@[simp] theorem to_nat_ordinal_inj {a b} : to_nat_ordinal a = to_nat_ordinal b ↔ a = b := iff.rfl
+@[simp] theorem of_nat_ordinal_inj {a b} : of_nat_ordinal a = of_nat_ordinal b ↔ a = b := iff.rfl
+
+theorem lt_to_nat_ordinal_iff {a b} : a < to_nat_ordinal b ↔ of_nat_ordinal a < b := iff.rfl
+theorem le_to_nat_ordinal_iff {a b} : a ≤ to_nat_ordinal b ↔ of_nat_ordinal a ≤ b := iff.rfl
+theorem eq_to_nat_ordinal_iff {a b} : a = to_nat_ordinal b ↔ of_nat_ordinal a = b := iff.rfl
+theorem lt_of_nat_ordinal_iff {a b} : a < of_nat_ordinal b ↔ to_nat_ordinal a < b := iff.rfl
+theorem le_of_nat_ordinal_iff {a b} : a ≤ of_nat_ordinal b ↔ to_nat_ordinal a ≤ b := iff.rfl
+theorem eq_of_nat_ordinal_iff {a b} : a = of_nat_ordinal b ↔ to_nat_ordinal a = b := iff.rfl
+
+@[simp] theorem to_nat_ordinal_max {a b} :
+  to_nat_ordinal (max a b) = max (to_nat_ordinal a) (to_nat_ordinal b) := rfl
+@[simp] theorem of_nat_ordinal_max {a b} :
+  of_nat_ordinal (max a b) = max (of_nat_ordinal a) (of_nat_ordinal b) := rfl
+@[simp] theorem to_nat_ordinal_min {a b} :
+  to_nat_ordinal (linear_order.min a b) = min (to_nat_ordinal a) (to_nat_ordinal b) := rfl
+@[simp] theorem of_nat_ordinal_min {a b} :
+  of_nat_ordinal (linear_order.min a b) = min (of_nat_ordinal a) (of_nat_ordinal b) := rfl
+
+/-- A recursor for `nat_ordinal`. Use as `induction x using nat_ordinal.rec`. -/
+protected def rec {β : nat_ordinal → Sort*} (h : Π a, β (to_nat_ordinal a)) : Π a, β a :=
+λ a, h (of_nat_ordinal a)
+
+/-- `ordinal.induction` but for `nat_ordinal`. -/
+theorem induction {p : nat_ordinal → Prop} :
+  ∀ i (h : ∀ j, (∀ k, k < j → p k) → p j), p i := ordinal.induction
+
+/-- `ordinal.blsub` but for `nat_ordinal`. -/
+def blsub (o : nat_ordinal.{u}) (f : Π a < o, nat_ordinal.{max u v}) : nat_ordinal :=
```

Comment:
You should add a rfl lemma for `blsub (to_nat_ordinal a)`

Reply:
As in, a lemma that unfolds the definition?

Reply:
No, a bit more than that. One that also cleans up a `to_nat_ordinal`:
```lean
lemma blsub_to_nat_ordinal (o : ordinal.{u}) (f : Π a < to_nat_ordinal o, nat_ordinal.{max u v}) :
  blsub (to_nat_ordinal o) f = to_nat_ordinal (o.blsub $ λ a ha, of_nat_ordinal (f (to_nat_ordinal a) ha)) :=
```

Reply:
I ended up removing this definition.